py/formak/python.py

```diff
@@ -376,6 +415,213 @@ def sensor_model(self, sensor_key, state, covariance, sensor_reading):
 
         return next_state, next_covariance
 
+    ### scikit-learn / sklearn interface ###
+
+    def _flatten_scoring_params(self, params):
+        # "process_noise": np.eye(1),
+        # "sensor_models": {"simple": {ui.Symbol("v"): ui.Symbol("v")}},
+        # "sensor_noises": {"simple": np.eye(1)},
+        flattened = list(np.diagonal(params["process_noise"]))
+        for key in sorted(list(params["sensor_models"])):
+            flattened.extend(np.diagonal(params["sensor_noises"][key]))
+
+        return flattened
+
+    def _inverse_flatten_scoring_params(self, flattened):
+        params = {k: v for k, v in self.params.items()}
+
+        controls, flattened = (
+            flattened[: self.control_size],
+            flattened[self.control_size :],
+        )
+
+        np.fill_diagonal(params["process_noise"], controls)
+
+        for key in sorted(list(self.params["sensor_models"])):
+            sensor_size = len(np.diagonal(params["sensor_noises"][key]))
+            sensor, flattened = flattened[:sensor_size], flattened[sensor_size:]
+            np.fill_diagonal(params["sensor_noises"][key], sensor)
+
+        return params
+
+    # Fit the model to data
+    def fit(self, X, y=None, sample_weight=None):
+        assert self.params["process_noise"] is not None
+        assert self.params["sensor_models"] is not None
+        assert self.params["sensor_noises"] is not None
+
+        x0 = self._flatten_scoring_params(self.params)
+
+        def minimize_this(x):
+            holdout_params = dict(self.get_params())
+
+            scoring_params = self._inverse_flatten_scoring_params(x)
+            self.set_params(**scoring_params)
+
+            score = self.score(X, y, sample_weight)
+
+            self.set_params(**holdout_params)
+            return score
+
+        minimize_this(x0)
+
+        result = minimize(minimize_this, x0)
+
+        if not result.success:
+            print("success", result.success, result.message)
+            assert result.success
+
+        soln_as_params = self._inverse_flatten_scoring_params(result.x)
+        self.set_params(**soln_as_params)
+
+        return self
+
+    # Compute the squared Mahalanobis distances of given observations.
+    def mahalanobis(self, X):
+        innovations, states, covariances = self.transform(X, include_states=True)
+        n_samples, n_sensors = innovations.shape
+
+        innovations = np.array(innovations).reshape((n_samples, n_sensors, 1))
+
+        return innovations.flatten()
+
+    # Compute the log-likelihood of X_test under the estimated Gaussian model.
```

Comment:
This comment isn't correct, I'm not doing anything with log likelihood, but the score still seems like it's useful